### DISCUSSION TITLE
Best way to consume Eclipse artifacts via Maven central (or other repository managers)

### DISCUSSION THREAD

**Discussion by vogella** [eclipse-tycho/tycho#2754] [Q&A]
IIRC @ptziegler wanted to consume Eclipse artifacts via a Maven repository manager like Artifactory in a Tycho build.

At this time, you could not use aggregated elements like features to consume the Eclipse artifacts via your target platform but had to list them individual in your target platform.

Is this still the case? Pure Maven artifacts describe their dependencies, maybe the feature artifact created by Tycho could contain this meta-data? 

I have a Tycho build which creates 60 plug-ins and uploads them to Artifactory. Afterwards multiple other Tycho builds are supposed to consume these artifacts. Maintaining the target platform for these consuming builds is a lot of effort.

**Comment by laeubi** [comment]
You can use `pom` in target and you can use `feature` in targets.

Nerveless your maven-dependencies should already declare their dependencies in the pom of that bundle so adding the dependency together with the option to include dependencies should be enough.

Beside that you can also deploy a maven-p2 site to artifactory and consume it like a regualr P2 site with the planner mode:
https://github.com/eclipse-tycho/tycho/blob/master/RELEASE_NOTES.md#creating-maven-p2-sites-with-tycho-packaging

**Reply by vogella** [reply]
Thanks @laeubi. It now noticed that the JAR files for plug-ins generated by Tycho contains their dependencies in the generated pom files. That help already a lot.

It looks like this meta-data is not generated for the JAR files for features. If this information could also be added to the feature JAR we could use the feature JAR from the Maven repo to define the target platform. The current workaround for this seems to be to have one additional plug-in listing all the plug-ins contained the feature as manifest dependencies and using this one in the target file.

**Comment by ptziegler** [comment]
> At this time, you could not use aggregated elements like features to consume the Eclipse artifacts via your target platform but had to list them individual in your target platform.

This was added quite some time ago with 6f4087698991133120949e83ed3c42724d789189 and its rather unsavory commit message and something that we still use internally. 😅 

> Is this still the case? Pure Maven artifacts describe their dependencies, maybe the feature artifact created by Tycho could contain this meta-data?

If all your plugins are contained by a feature then this should already work, as long as they are part of the same reactor build. Though I _think_ you have to explicitly enable the pom generation. But I'm not 100% sure anymore.

```
<plugin>
	<groupId>org.eclipse.tycho</groupId>
	<artifactId>tycho-packaging-plugin</artifactId>
	<configuration>
		<skipPomGeneration>false</skipPomGeneration>
		<archive>
			<addMavenDescriptor>false</addMavenDescriptor>
		</archive>
	</configuration>
</plugin>
```

Creating a Maven p2 site sounds like a really interesting concept. And in general, I think this is the preferrable solution for this type of problem.

**Reply by vogella** [reply]
Thanks @ptziegler I saw your comment a little bit too late. [6f4087698991133120949e83ed3c42724d789189](https://github.com/eclipse-tycho/tycho/commit/6f4087698991133120949e83ed3c42724d789189) sounds like the desired behavior.  I have to test this in this new environment, but it sounds very promising. Maybe you could add it to https://github.com/eclipse-tycho/tycho/blob/master/RELEASE_NOTES.md (if not yet done, but I have not found it there) so that is easier to discover.

**Reply by vogella** [reply]
@ptziegler did you do something similar for PDE or it is currently only a solution for Tycho? If it is currently only for Tycho, how do you handle it in the IDE?

**Reply by laeubi** [reply]
You could simply try it out ;-) Beside that the target support is part of m2e ...

**Reply by ptziegler** [reply]
> @ptziegler did you do something similar for PDE or it is currently only a solution for Tycho? If it is currently only for Tycho, how do you handle it in the IDE?

m2e supports this [too](https://github.com/eclipse-m2e/m2e-core/pull/636). Otherwise it would try to rebundle the feature as an OSGi bundle. But even then: As long as the Maven metadata are valid, m2e should also fetch all dependent plugins (assuming that the dependency depth isn't set to "none").

Though the reason we needed it for Tycho is because our build consists of 5 separate phases, where each phase requires the artifacts from its previous phase. And sadly Nexus doesn't support hosted p2 repositories, so we had to go the Maven route.
Inside the IDE, we have everything checked out together, so we don't need to add the feature to our "core" target platform definition.

**Reply by vogella** [reply]
Just tested a Feature dependency from Maven  and it works great with all its dependencies.

**Comment by vogella** [comment]
> Creating a Maven p2 site sounds like a really interesting concept. And in general, I think this is the preferrable solution for this type of problem.

Tried this also. It works also great, only the JGit timestamp provider is not used. Is their a separate configuration need for this?
```
 <plugin>
          <groupId>org.eclipse.tycho</groupId>
          <artifactId>tycho-packaging-plugin</artifactId>
          <version>${tycho.version}</version>
        
         <dependencies>
            <dependency>
              <groupId>org.eclipse.tycho</groupId>
              <artifactId>tycho-buildtimestamp-jgit</artifactId>
              <version>${tycho.version}</version>
            </dependency>
            <dependency>
              <groupId>org.eclipse.tycho.extras</groupId>
              <artifactId>tycho-sourceref-jgit</artifactId>
              <version>${tycho.version}</version>
            </dependency>
          </dependencies>

          <configuration>
            <timestampProvider>jgit</timestampProvider>
            <jgit.dirtyWorkingTree>warning</jgit.dirtyWorkingTree>
           
          </configuration>
        </plugin>
```

**Reply by laeubi** [reply]
What do you mean by "JGit timestamp provider is not used"?
